source/utilities.tex

```diff
@@ -17022,6 +17122,69 @@
 
+\pnum
+For the \tcode{common_reference} trait applied to a parameter pack \tcode{T} of
+types, the member \tcode{type} shall be either defined or not present as follows:
+\begin{itemize}
+\item If \tcode{sizeof...(T)} is zero, there shall be no member \tcode{type}.
+
+\item Otherwise, if \tcode{sizeof...(T)} is one, let \tcode{T0} denote the sole
+  type in the pack \tcode{T}. The member typedef \tcode{type} shall denote the
+  same type as \tcode{T0}.
+
+\item Otherwise, if \tcode{sizeof...(T)} is two, let \tcode{T1} and \tcode{T2}
+  denote the two types in the pack \tcode{T}. Then
+  \begin{itemize}
+  \item If \tcode{T1} and \tcode{T2} are reference types and
+    \tcode{COMMON_REF(T1, T2)} is well-formed, then the member typedef
+    \tcode{type} denotes that type.
+
+  \item Otherwise, if \tcode{basic_common_reference<remove_cvref_t<T1>, remove_cvref_t<T2>,
+    \brk{}XREF(\brk{}T1), XREF(T2)>::type} is well-formed, then the member typedef
+    \tcode{type} denotes that type.
+
+  \item Otherwise, if \tcode{COND_RES(T1, T2)} is well-formed, then the
+    member typedef \tcode{type} denotes that type.
+
+  \item Otherwise, if \tcode{common_type_t<T1, T2>} is well-formed, then the
+    member typedef \tcode{type} denotes that type.
+
+  \item Otherwise, there shall be no member \tcode{type}.
+  \end{itemize}
+
+\item Otherwise, if \tcode{sizeof...(T)} is greater than two, let \tcode{T1},
+  \tcode{T2}, and \tcode{Rest}, respectively, denote the first, second, and
+  (pack of) remaining types comprising \tcode{T}. Let \tcode{C} be the type
+  \tcode{common_reference_t<T1, T2>}. Then:
+  \begin{itemize}
+  \item If there is such a type \tcode{C}, the member typedef \tcode{type} shall
+    denote the same type, if any, as \tcode{common_reference_t<C, Rest...>}.
+
+  \item Otherwise, there shall be no member \tcode{type}.
+  \end{itemize}
+\end{itemize}
+
+\pnum
+Notwithstanding the provisions of\iref{meta.type.synop}, and
+pursuant to\iref{namespace.std}, a program may specialize
+\tcode{basic_common_reference<T, U, TQual, UQual>}
+for types \tcode{T} and \tcode{U} such that
+\tcode{is_same_v<T, decay_t<T>{>}} and
+\tcode{is_same_v<U, decay_t<U>{>}} are each \tcode{true}.
+\begin{note}
+Such specializations are needed when only explicit conversions are desired between
+the template arguments.
+\end{note}
+Such a specialization need not have a member named \tcode{type}, but if it does,
+that member shall be a \grammarterm{typedef-name} for an accessible and
+unambiguous type \tcode{C} to which each of the types \tcode{TQual<T>} and
+\tcode{UQual<U>} is convertible.
+Moreover, \tcode{basic_common_reference<T, U, TQual, UQual>::type} shall denote
+the same type, if any, as does \tcode{basic_common_reference<U, T, UQual, TQual>::type}.
+A program shall not specialize \tcode{basic_common_reference} on the third or
+fourth parameters, \tcode{TQual} or \tcode{UQual}. No diagnostic is required for
```

Comment:
What does that mean? Are you talking about partial specialization here? Then you should say so at the start of this paragraph. (A specialization is a non-template; a partial specialization is a template: two fundamentally different things.)

Reply:
Yes, the intent here is that users partially specialize `basic_common_reference`. I admit that I can't see a reason now for "A program shall not specialize `\tcode{basic_common_reference}` on the third or fourth parameters, `\tcode{TQual}` or `\tcode{UQual}`." I believe the intent is to make it harder for users to do something stupid and violate the other requirements and/or point them more directly down the intended path.

@ericniebler, can you clarify the intent of this particular requirement?

Reply:
Yes, the intent here is to disallow specifying the third or fourth template parameters in either a partial or explicit specialization. Those parameters are specified by the `common_reference` trait. The exact unary templates used as template arguments to the (possibly user-specialized) `basic_common_reference` instantiation are unspecified, and implementors are free to change them between releases. Any code that specializes on those parameters (either partial or explicit) is _de facto_ wrong.

Reply:
Does a program with such a specialization *need* to be IFNDR? IMO it's sufficient that `common_reference` specifies how and when it works, and users shouldn't be surprised when their broken specializations are never selected.

Reply:
I suppose not. Striketh sentence if it offendeth thee.

An alternative which I hadn't considered until just now is to redesign the trait to make it impossible to misuse:

```c++
template <class A, class B>
struct basic_common_reference {
  template <template <class> class AQual, template <class> class BQual>
  using type = ...;
};
```

Reply:
I don't know that I buy "impossible to misuse", but I'll make a note to possibly file an issue to make this design change after we investigate it more thoroughly.

...and strike the requirement for now, since we think it adds no useful normative effect.

Reply:
I did change the first sentence of what is now "Note D" to read:
> Notwithstanding the provisions of \ref{meta.type.synop}, and pursuant to \ref{namespace.std}, a program may **partially** specialize \tcode{basic_common_reference<T, U, TQual, UQual>} for types \tcode{T} and \tcode{U} such that \tcode{is_same_v<T, decay_t<T>{>}} and \tcode{is_same_v<U, decay_t<U>{>}} are each \tcode{true}.

which I consider to be an editorial clarification. If you feel that change is insufficient or not editorial, I'll revert it and file an issue.